enterprise/server/backends/migration_cache/migration_cache.go

```diff
@@ -146,31 +148,174 @@ func pebbleCacheFromConfig(env environment.Env, cfg *PebbleCacheConfig) (*pebble
+		})
+	}
+
+	if err := eg.Wait(); err != nil {
+		return false, err
+	}
+
+	if dstErr != nil {
+		log.Warningf("Migration dest %v contains failed: %s", d, dstErr)
+	} else if mc.logNotFoundErrors && srcContains != dstContains {
+		log.Warningf("Migration digest %v src contains %v, dest contains %v", d, srcContains, dstContains)
+	}
+
+	return srcContains, srcErr
 }
 
 func (mc *MigrationCache) Metadata(ctx context.Context, d *repb.Digest) (*interfaces.CacheMetadata, error) {
-	return nil, status.UnimplementedError("not yet implemented")
+	eg, gctx := errgroup.WithContext(ctx)
+	var srcErr, dstErr error
+	var srcMetadata *interfaces.CacheMetadata
+
+	eg.Go(func() error {
+		srcMetadata, srcErr = mc.src.Metadata(gctx, d)
+		return srcErr
+	})
+
+	doubleRead := rand.Float64() <= mc.doubleReadPercentage
+	if doubleRead {
+		eg.Go(func() error {
+			_, dstErr = mc.dest.Metadata(gctx, d)
+			return nil // we don't care about the return error from this cache
+		})
+	}
+
+	if err := eg.Wait(); err != nil {
+		return nil, err
+	}
+
+	if dstErr != nil && (mc.logNotFoundErrors || !status.IsNotFoundError(dstErr)) {
+		log.Warningf("Migration dest %v metadata failed: %s", d, dstErr)
+	}
+
+	return srcMetadata, srcErr
 }
 
 func (mc *MigrationCache) FindMissing(ctx context.Context, digests []*repb.Digest) ([]*repb.Digest, error) {
-	return nil, status.UnimplementedError("not yet implemented")
+	eg, gctx := errgroup.WithContext(ctx)
+	var srcErr, dstErr error
+	var srcMissing []*repb.Digest
+
+	eg.Go(func() error {
+		srcMissing, srcErr = mc.src.FindMissing(gctx, digests)
+		return srcErr
+	})
+
+	doubleRead := rand.Float64() <= mc.doubleReadPercentage
+	if doubleRead {
+		eg.Go(func() error {
+			_, dstErr = mc.dest.FindMissing(gctx, digests)
+			return nil // we don't care about the return error from this cache
+		})
+	}
+
+	if err := eg.Wait(); err != nil {
+		return nil, err
+	}
+
+	if dstErr != nil {
+		log.Warningf("Migration dest FindMissing %v failed: %s", digests, dstErr)
+	}
+
+	return srcMissing, srcErr
 }
 
 func (mc *MigrationCache) GetMulti(ctx context.Context, digests []*repb.Digest) (map[*repb.Digest][]byte, error) {
-	return nil, status.UnimplementedError("not yet implemented")
+	lock := sync.RWMutex{} // protects(foundMap)
+	foundMap := make(map[*repb.Digest][]byte, len(digests))
+	eg, ctx := errgroup.WithContext(ctx)
+
+	for _, d := range digests {
+		fetchFn := func(d *repb.Digest) {
+			eg.Go(func() error {
+				data, err := mc.Get(ctx, d)
+				if status.IsNotFoundError(err) {
+					return nil
+				}
+				if err != nil {
+					return err
+				}
+				lock.Lock()
+				foundMap[d] = data
+				lock.Unlock()
+				return nil
+			})
+		}
+		fetchFn(d)
+	}
+
+	if err := eg.Wait(); err != nil {
+		return nil, err
+	}
+
+	return foundMap, nil
 }
 
 func (mc *MigrationCache) SetMulti(ctx context.Context, kvs map[*repb.Digest][]byte) error {
-	return status.UnimplementedError("not yet implemented")
+	for d, data := range kvs {
```

Comment:
Similarly to GetMulti, we ideally should be calling SetMulti on the underlying caches (though in practice the underlying implementations call Set)

Reply:
Done